File: 6kyu/Simple string indices/C++/src/solution.cpp
// This is an independent project of an individual developer. Dear PVS-Studio, please check it.
// PVS-Studio Static Code Analyzer for C, C++ and C#: http://www.viva64.com
#include "solution.hpp"

#include <gsl-lite.hpp>

#include <limits>

using gsl::narrow;
using std::numeric_limits;

int solve(const std::string& str, size_t index)
{
	const char openBrace = '(';
	const char closeBrace = ')';
	const int error = -1;
	if (str.at(index) != openBrace) {
		return error;
	}
	size_t numberOpenBraces = 1;
	for (size_t i = index + 1ULL; i < str.size(); ++i) {
		switch (str.at(i)) {
		case openBrace: 
			++numberOpenBraces;
			break;
		case closeBrace:
			--numberOpenBraces;
			break;
		default:
			/* do nothing */
			break;
		}
		if (numberOpenBraces == 0) {
			if (i > narrow<size_t>(numeric_limits<int>::max())) {
				return error;
			}
			return narrow<int>(i);
		}
	}
	return error;
}

File: 6kyu/Simple string indices/Typescript/src/solution.ts
export const solution = (str: string, index: number): number => {
    const openBrace = "(";
    const closeBrace = ")";
    const error = -1;

    if (str.charAt(index) !== openBrace) {
        return error;
    }

    let numberOpenBraces = 1;
    for (let i = index + 1; i < str.length; ++i) {
        switch (str.charAt(i)) {
            case openBrace: {
                ++numberOpenBraces;
                break;
            }
            case closeBrace: {
                --numberOpenBraces;
                break;
            }
            default: {
                // Do nothing
            }
        }

        if (numberOpenBraces === 0) {
            return i;
        }
    }

    return error;
};
